src/mjolnir/graphenhancer.cc

```diff
@@ -94,58 +96,232 @@ struct enhancer_stats {
   }
 };
 
+/*
+The json basically looks like this:
+
+[{
+  "iso3166-1": "us",
+  "iso3166-2": "pa",
+  "urban": {
+    "way": [1,2,3,4,5,6,7,8],
+    "link_exiting": [9,10,11,12,13,14],
+    "link_turning": [15,16,17,18,19,20],
+    "roundabout": [21,22,23,24,25,26,27,28],
+    "driveway": 29,
+    "alley": 30,
+    "parking_aisle": 31,
+    "drive-through": 32
+  },
+  "rural": {
+    "way": [33,34,35,36,37,38,39,40],
+    "link_exiting": [41,42,43,44,45,46],
+    "link_turning": [47,48,49,50,51,52],
+    "roundabout": [53,54,55,56,57,58,59,60],
+    "driveway": 61,
+    "alley": 62,
+    "parking_aisle": 63,
+    "drive-through": 64
+  }
+}]
+ */
+
 /**
- * Update directed edge speed based on density and other edge parameters like
- * surface type. TODO - add admin specific logic
- * @param  directededge  Directed edge to update.
- * @param  density       Relative road density.
- * @param  urban_rc_speed Array of default speeds vs. road class for urban areas
- * @param  rc_speed Array of default speeds vs. road class
- * @param  infer_turn_channels flag indicating if we should infer tc.
  *
  */
-void UpdateSpeed(DirectedEdge& directededge,
-                 const uint32_t density,
-                 const uint32_t* urban_rc_speed,
-                 bool infer_turn_channels) {
-
-  // Update speed on ramps (if not a tagged speed) and turn channels
-  if (directededge.link()) {
-    uint32_t speed = directededge.speed();
-    Use use = directededge.use();
-    if (use == Use::kTurnChannel && infer_turn_channels) {
-      speed = static_cast<uint32_t>((speed * kTurnChannelFactor) + 0.5f);
-    } else if ((use == Use::kRamp) && (directededge.speed_type() != SpeedType::kTagged)) {
-      // If no tagged speed set ramp speed to slightly lower than speed
-      // for roads of this classification
-      RoadClass rc = directededge.classification();
-      if ((rc == RoadClass::kMotorway) || (rc == RoadClass::kTrunk) || (rc == RoadClass::kPrimary)) {
-        speed = (density > 8) ? static_cast<uint32_t>((speed * kRampDensityFactor) + 0.5f)
-                              : static_cast<uint32_t>((speed * kRampFactor) + 0.5f);
-      } else {
-        speed = static_cast<uint32_t>((speed * kRampFactor) + 0.5f);
+struct SpeedAssigner {
+  struct SpeedTable {
```

Comment:
this object closely mirrors the json but is easier to use than a rapidjson object so we parse the json into it